File: controller/channel_inflight.go
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
package controller

import (
	"net/http"
	"strconv"

	"github.com/QuantumNous/new-api/common"
	"github.com/QuantumNous/new-api/i18n"
	"github.com/QuantumNous/new-api/relay/inflight"

	"github.com/gin-gonic/gin"
)

// All three handlers answer purely from the in-process registry. They deliberately
// do not verify that the channel exists: the admin console polls them on its
// refresh beat, and a database round trip per poll would turn an observability
// panel into load. An unknown id simply reports an empty channel.

// GetChannelInflight reports what one channel currently has in flight on this
// instance, without touching a single connection. It backs the confirmation
// dialog, which needs a reading fresher than the table's polling beat.
func GetChannelInflight(c *gin.Context) {
	channelId, ok := channelIdParam(c)
	if !ok {
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "",
		"data":    inflight.Preview(channelId),
	})
}

// GetChannelInflightRuntime reports every channel that has something in flight on
// this instance. The channel table polls this one endpoint on its existing refresh
// beat; a per-row endpoint would multiply the poll by the number of visible rows.
func GetChannelInflightRuntime(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "",
		"data":    inflight.CurrentRuntime(),
	})
}

// CleanupChannelInflight tears down the connections of one channel that have
// already outlived the gateway's own timeout contract. Connections still inside
// their contract are never touched, so a healthy channel reports zero cleared.
func CleanupChannelInflight(c *gin.Context) {
	channelId, ok := channelIdParam(c)
	if !ok {
		return
	}
	snapshot := inflight.Cleanup(channelId)
	if snapshot.Cancelled > 0 {
		recordManageAudit(c, "channel.inflight_cleanup", map[string]interface{}{
			"id":        channelId,
			"count":     snapshot.Cancelled,
			"in_flight": snapshot.InFlight,
			"by_reason": snapshot.ByReason,
		})
	}
	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "",
		"data":    snapshot,
	})
}

func channelIdParam(c *gin.Context) (int, bool) {
	channelId, err := strconv.Atoi(c.Param("id"))
	if err != nil || channelId <= 0 {
		common.ApiErrorI18n(c, i18n.MsgInvalidParams)
		return 0, false
	}
	return channelId, true
}

File: web/classic/src/helpers/channel-inflight.js
/*
Copyright (C) 2025 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/

const toNonNegativeCount = (value) => {
  const number = Number(value);
  return Number.isFinite(number) && number > 0 ? number : 0;
};

/** Age of the oldest live connection, rendered compactly enough for a table cell. */
export function formatInflightAge(ms) {
  const value = Number(ms);
  if (!Number.isFinite(value) || value <= 0) return '0s';
  const seconds = Math.floor(value / 1000);
  if (seconds < 60) return `${seconds}s`;
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes}m${seconds % 60}s`;
  return `${Math.floor(minutes / 60)}h${minutes % 60}m`;
}

/**
 * Connections that have not produced a first byte yet: the upstream accepted the
 * request but has answered with neither response headers nor a first chunk. This
 * is the number operations describes as "长连接卡住".
 */
export function countAwaitingFirstByte(inflight) {
  if (!inflight || typeof inflight !== 'object') return 0;
  return (
    toNonNegativeCount(inflight.awaiting_headers) +
    toNonNegativeCount(inflight.awaiting_first_chunk)
  );
}

/**
 * Index the bulk runtime response by channel id so a row can look itself up.
 * One request per poll covers every visible row; never poll this per row.
 */
export function indexInflightByChannel(runtime) {
  const indexed = {};
  const channels = Array.isArray(runtime?.channels) ? runtime.channels : [];
  for (const channel of channels) {
    if (!channel || typeof channel !== 'object') continue;
    const id = Number(channel.channel_id);
    if (!Number.isFinite(id)) continue;
    indexed[String(id)] = channel;
  }
  return indexed;
}

/**
 * The per-class timeout contract, rendered for the confirmation dialog.
 * A class whose timeout is 0 has no overdue threshold, so its connections can
 * never become clearable — say so instead of printing a misleading "0s".
 */
export function formatInflightThreshold(seconds, enabled, notConfiguredText) {
  return enabled ? `${toNonNegativeCount(seconds)}s` : notConfiguredText;
}
